### PR TITLE
fix#283240 Let Style dlg state persist through open/close

### PR DESCRIPTION
**Pull request by L0uisc**
This PR fixes https://musescore.org/en/node/283240 by adding an `EditStyle*` pointer to the `MuseScore` class and only creating it once. This allows the previous state to be "remembered" when a user opens the dialog again. It also should make subsequent opening of the dialog faster. I also changed the call to the `exec()` method to `open()` as per https://doc.qt.io/qt-5/qdialog.html#exec

### CODE REVIEW

**Comment by dmitrio95** [comment]
> I also changed the call to the `exec()` method to `open()` as per https://doc.qt.io/qt-5/qdialog.html#exec

It looks like this change can result in unexpected bugs in this case. For example, see the following scenario (tested on Linux, not sure whether it works on other platforms):

1) Open any score
2) Select a note
3) Open Style dialog
4) Click outside the Style dialog
5) Press Up arrow.
First unexpected result: the selected note actually changes its pitch
6) In the opened Style dialog change a musical symbols font (or any other setting)
7) Press OK
8) Try to undo the changes (Ctrl+Z)
Second unexpected result: font change is not undoable.

So it looks like changing `exec()` to `open()` will require more changes throughout MuseScore code in order to make it work properly, at least regarding keyboard events handling.

**Comment by L0uisc** [comment]
I can reproduce the second unexpected behaviour on Windows 7 32 bit, but not the first. I guess it is because the result of the async call is never polled, so it is never added to the undo stack. Probably the easiest to do is to just revert to using `exec()`

**Comment by L0uisc** [comment]
I can still reproduce if I change `open()` back to `exec()`, so it must be another change causing the second thing you see. I am building current master to see if it is the other code in this commit or some totally unrelated code causing it.

**Comment by L0uisc** [comment]
@dmitrio95 it is my changes, but not the `open()` call, which is causing the inability to undo.

**Comment by L0uisc** [comment]
Another thing I realize now is that there is no way to change on which score the dialog acts under this PR. It is always affecting the score which was current at the time it was constructed. So maybe I should just go back to the caching in a static member approach.

**Comment by dmitrio95** [comment]
> I can reproduce the second unexpected behaviour on Windows 7 32 bit, but not the first.

In this case I suspect what you see may be just not updating the Style dialog controls when it gets reopened (if you check the values in the Style dialog instead of the score appearance). This happens because all the displayed values are set in the `EditStyle`'s constructor which doesn't get called on the dialog reopening with these changes. This indeed does not depend on calling `open()` or `exec()`.

What I report may be related to the difference in the dialog's modality (window-modal dialog in case of `open()` vs. application-modal dialog in case of `exec()`) but that needs checking.

**Comment by L0uisc** [comment]
@dmitrio95 I need to move some of the code in the constructor into the showEvent() method so the dialog's edit fields, dropdowns, etc. show an up to date version of the score info. The other option is to keep the code in musescore.cpp as it is and just adding a static member to the EditStyle class, which caches the last shown page. On construction, the current page is then set to the cached value.

**Comment by L0uisc** [comment]
This should be everything. Comments welcome!

**Comment by L0uisc** [comment]
@dmitrio95 can you figure out where to put the `EditStyle*` in `MuseScore`? The style dialog is shown in one [other place](https://github.com/musescore/MuseScore/blob/master/mscore/propertymenu.cpp#L466), and currently I just kept it as a local variable and let it be constructed each time those lines are executed. Since my code now has a pointer to a heap-allocated `EditStyle`, I want to try to use only the one object and reference it from both places. The code as it is in this PR works, I just want to try to unify the access even more.

The other thing we have to keep in mind, though, is that if I can unify those two calls into one object, any subsequent opening of the style dialog will open on the "Notes" page, while as it is currently, the previous explicit page is remembered. Something to think about.

**Comment by IsaacWeiss** [comment]
I don't think that other call makes sense at all. Why should a link to the relevant page of the Style dialog be provided for the context menu of noteheads, but no other elements?

The best approach to this discrepancy might be to make that command available for more elements, always going their respective pages, but tbh I'd just get rid of it for notes. No one will miss it, and your immediate problem is solved.

**Comment by L0uisc** [comment]
@dmitrio95 I think this should be everything. I solved the problem I had previously about the call in [propertymenu.cpp](https://github.com/musescore/MuseScore/blob/master/mscore/propertymenu.cpp#L466), which is more important now that #5093 exists.

**Comment by dmitrio95** [comment]
The code looks and performs well but I am not yet sure about the idea of having separate style dialog instances per score view. First, the style dialog is relatively heavy and I am not sure how it will perform in terms of memory when having a large number of scores opened (but maybe there won't be any issues here). Second, when having a side-by-side view enabled this solution may perform not fully as expected: the style dialog instance being shown will depend on which side of a side-by-side view does currently have a focus. What do you think on this? Could these issues be bad, and is it desirable to have a separate style dialog state per score or per application?

**Comment by L0uisc** [comment]
Yes, well, that case of two instances of the same score open side-by-side is pretty bad. I just didn't find a way to let `ScoreView` access the same dialog if it was declared in `MuseScore`. Do you think it would be a good idea to add a private `EditStyle*` in `ScoreView` and setting it to the address of `MuseScore`'s `EditStyle` on construction? I don't know if `MuseScore` or the `ScoreView`s are constructed first, but I guess `MuseScore` should be first.

**Comment by L0uisc** [comment]
@dmitrio95 so I am starting to think the best would be to have a global `EditStyle*` which could be accessed from both `MuseScore` and `ScoreView`. Any thoughts? The alternative, which is giving me problems, is to try and synchronize an `EditStyle*` in `MuseScore` and a `static EditStyle*` in `ScoreView`.

**Comment by dmitrio95** [comment]
Actually if one `EditStyle` object per application is sufficient it should be fine to store it inside MuseScore. There is a global `mscore` variable pointing to a MuseScore instance is available, and it can be used (and is already used in many places) to access that dialog from within `ScoreView`. Not sure whether it is a really good practice but it is widely used in MuseScore code so there shouldn't be anything bad in applying it in this case too.

**Comment by L0uisc** [comment]
Thanks for the tip about the `mscore` object! I think this should then solve the problem of multiple dialogs and the problem of two call sites constructing different objects.

File: mscore/scoreview.h [hunk 1]
```diff
@@ -88,6 +89,7 @@ class ScoreView : public QWidget, public MuseScoreView {
 
       ViewState state;
       OmrView* _omrView;
+      static EditStyle* styleDlg;
```
Comment (dmitrio95) [inline mscore/scoreview.h]: This is not needed anymore, is it?

File: mscore/scoreview.h [hunk 2]
```diff
@@ -46,6 +46,7 @@ class Tuplet;
 class FretDiagram;
 class Bend;
 class TremoloBar;
+class EditStyle;
```
Comment (dmitrio95) [inline mscore/scoreview.h]: ...and this declaration too?

Comment (L0uisc) [reply mscore/scoreview.h]: Indeed, updated.

File: mscore/musescore.cpp [hunk 3]
```diff
@@ -6134,8 +6134,9 @@ void MuseScore::cmd(QAction* a, const QString& cmd)
                   }
             }
       else if (cmd == "edit-style") {
-            EditStyle es(cs, this);
-            es.exec();
+            if (!_styleDlg)
+                  _styleDlg = new EditStyle { cs, this };
+            _styleDlg->exec();
```
Comment (dmitrio95) [inline mscore/musescore.cpp]: Looking at this once more, it seems you have missed the `setScore()` call before executing the dialog (both here and in `ScoreView`). This makes style dialog edit a style of a wrong score, at least after it was created once. Could you please add the missing calls so I could merge this pull request?

Comment (L0uisc) [reply mscore/musescore.cpp]: Done

**Comment by dmitrio95** [comment]
This looks good now but there seems to be some issue in my tests. Try the following steps:
1) Open two scores (with style settings that are close to default).
2) For the first score open style dialog, switch to "Measure Numbers" tab.
3) Switch from "Every System" to "Interval",  set the interval value to 1. Press OK.
4) Switch to the second score, open and close the style dialog.
5) Switch back to the first score, open the style dialog.

Result: the dialog shows the "Every System" setting chosen for measure numbers. The interval value is still correctly set to 1, and the score seems to correctly follow the previously chosen settings.

In current master this issue does not exist for me (and it is indeed likely to be related to updating values for an existing dialog).

I don't know currently what can be the cause of this issue but it can be misleading for users if this issue really exists. Is this issue reproducible for you?

**Comment by dmitrio95** [comment]
Also could you please add a whitespace to the commit message (`fix #283240` instead of `fix#283240`)? I suspect the bot on musescore.org may not recognize this pattern without a whitespace.

**Comment by L0uisc** [comment]
@dmitrio95 I can reproduce. I guess I have missed something that needs to be moved from the `EditStyle` constructor into the `showEvent()` event handler.

**Comment by L0uisc** [comment]
I figured out what the problem was: the radio button for selecting making measure numbers every x measures isn't in `styleWidgets`, so it isn't updated in the for loop in `EditStyle::setValues()`. I think more testing is in order, because there is a very good chance that there are more such cases. But here is the fix for this specific case in the meantime.

**Comment by L0uisc** [comment]
Obviously, I made a mistake, because not all my changes are pushed, but I see them all locally and `git status` doesn't show any untracked files with changes. Give me a moment to figure it out.

**Comment by MarcSabatella** [comment]
(originally posted in error about #5093 )

FWIW, I do like the idea behind this, and would be happy to see it merged. But one thing to be careful of is documentation. Right now I for one have been in the habit of telling people that, for instance, "Hide empty staves" is found on Format / Style, which worked because Score was always the default page. Now we need to be careful to always say Format / Style / Score. Might want to go through the Handbook and see if anything needs updating.

**Comment by L0uisc** [comment]
I currently have a bit of a problem with caching an instance of the dialog on the heap when you open it from multiple scores. Changing something in one score, then switching to another and opening the dialog will change its settings too. I think I need to either dig a lot into how Qt signals/slots work or I should just cache the page no in a static member and then set the correct last page after construction. I think I will like to do the first, but if I can't figure it out, the second should work fine.

So this PR will take some work from me to be mergeable anyway. But I will update the docs when we agree on the final form of the PR.

**Comment by MarcSabatella** [comment]
FWIW, I wouldn't be trying to manage any of that yourself.  I'd make this a private data member associated with the score, or the scoreview, or the scoretab, or whatever other higher-level construct you want it to change with.  Probably scoretab, I guess, but I haven't really looked.

**Comment by Jojo-Schmitz** [comment]
rebase needed

**Comment by dmitrio95** [comment]
Rebase is indeed needed.

Also because of this:

> I figured out what the problem was: the radio button for selecting making measure numbers every x measures isn't in `styleWidgets`, so it isn't updated in the for loop in `EditStyle::setValues()`. I think more testing is in order, because there is a very good chance that there are more such cases.

I believe it would be better to move it to 3.4 version milestone instead to allow it some time for testing in beta versions.

File: mscore/editstyle.cpp [hunk 4]
```diff
@@ -38,12 +38,11 @@ namespace Ms {
 //---------------------------------------------------------
 
 EditStyle::EditStyle(Score* s, QWidget* parent)
-   : QDialog(parent)
+   : cs(s), QDialog(parent)
```
Comment (Jojo-Schmitz) [inline mscore/editstyle.cpp]: This does cause a new compiler warning in MinGW:

In file included from C:\Users\Jojo\Documents\GitHub\MuseScore\mscore\editstyle.cpp:16:0:
C:\Users\Jojo\Documents\GitHub\MuseScore\mscore\editstyle.h: In constructor 'Ms::EditStyle::EditStyle(Ms::Score*, QWidget*)':
C:\Users\Jojo\Documents\GitHub\MuseScore\mscore\editstyle.h:59:14: warning: 'Ms::EditStyle::cs' will be initialized after [-Wreorder]
       Score* cs;
              ^~
C:\Users\Jojo\Documents\GitHub\MuseScore\mscore\editstyle.cpp:41:27: warning:   base 'QDialog' [-Wreorder]
    : cs(s), QDialog(parent)
                           ^
C:\Users\Jojo\Documents\GitHub\MuseScore\mscore\editstyle.cpp:40:1: warning:   when initialized here [-Wreorder]
 EditStyle::EditStyle(Score* s, QWidget* parent)

No idea though how to fix it

Comment (dmitrio95) [reply mscore/editstyle.cpp]: Just reorder QDialog constructor call and `cs` initialization:
`: QDialog(parent), cs(s)`

These two do not depend on each other so in this case this warning doesn't point to any real issue.

Comment (Jojo-Schmitz) [reply mscore/editstyle.cpp]: Thanks, see #5548

**Pull request merged**